src/components/file-browser/FileBrowserDirectories.ts

```diff
@@ -0,0 +1,369 @@
+import { Dispatch, Fragment, ReactNode, SetStateAction, useEffect, useRef, useState } from 'react'
+import { h, li, p, span, ul } from 'react-hyperscript-helpers'
+import { Link } from 'src/components/common'
+import { useDirectoriesInDirectory } from 'src/components/file-browser/file-browser-hooks'
+import { basename } from 'src/components/file-browser/file-browser-utils'
+import { icon } from 'src/components/icons'
+import Interactive, { InteractiveProps } from 'src/components/Interactive' // eslint-disable-line import/named
+import FileBrowserProvider from 'src/libs/ajax/file-browser-providers/FileBrowserProvider'
+import colors from 'src/libs/colors'
+import * as Utils from 'src/libs/utils'
+
+
+interface FileBrowserDirectoryContentsProps {
+  activeDescendant: string
+  level: number
+  parentId: string
+  path: string
+  provider: FileBrowserProvider
+  rootLabel: string
+  selectedDirectory: string
+  setActiveDescendant: Dispatch<SetStateAction<string>>
+  onFinishedLoading: () => void
+  onSelectDirectory: (path: string) => void
+}
+
+const renderDirectoryStatus = (level: number, children: ReactNode) => p({
+  style: {
+    // Match padding + border of tree items.
+    padding: 'calc(0.25rem + 1px) 0.5rem',
+    // Align with arrow icons of tree items.
+    margin: `0 0 0 ${level + 0.75}rem`
+  }
+}, [children])
+
+export const FileBrowserDirectoryContents = (props: FileBrowserDirectoryContentsProps) => {
+  const {
+    activeDescendant,
+    level,
+    parentId,
+    path,
+    provider,
+    rootLabel,
+    selectedDirectory,
+    setActiveDescendant,
+    onFinishedLoading,
+    onSelectDirectory
+  } = props
+
+  const directoryLabel = path === '' ? rootLabel : basename(path)
+
+  const loadedAlertElementRef = useRef<HTMLSpanElement | null>(null)
+
+  const {
+    state: { status, directories },
+    hasNextPage,
```

Comment:
Why are there no types on these?

Reply:
They are known from the return type of `useDirectoriesInDirectory`.

Reply:
This is something I'm trying to get used to… that the types aren't repeated so when you are reading the code outside of an IDE it's not easy to tell what the types are.